### PR TITLE
Validate V5 transactions with Sapling shielded data

### LINKED ISSUE #1981: Implement Transparent and Sapling validation for transaction v5
**Is your feature request related to a problem? Please describe.**

In #1829 and #1980, we parse and check transparent and sapling in transaction v5. But we need to call those functions in the transaction verifier.

**Describe the solution you'd like**

add `Transaction::V5` to `transaction::check` (#1980 + PRs, #2070):
- [x] quote the relevant consensus rules in each function's docs
- [x] add orchard TODOs
- [x] refactor to remove duplicate code, and make the consensus rule logic clearer
- [x] implement the missing "no `PrevOut`s" coinbase check, added a few days ago in https://github.com/zcash/zips/commit/10710d92a61b0090ed898dfa473cdc0e0cff81f3

add `Transaction::V5` to `transaction::Verifier`:
- [x] reject V5 transactions before the NU5 activation height on the configured network
- [x] make sure `Transaction.network_upgrade` matches the currently active network upgrade (#2100)
  - the current upgrade is based on the height and configured network, it won't always be NU5

Other transparent and sapling consensus rules:
- [x] quote the relevant consensus rules in each function's docs
- [x] add orchard TODOs
- [x] refactor to remove duplicate code, and make the consensus rule logic clearer

**Describe alternatives you've considered**

These consensus rules are required for NU5.

**Additional context**

We also need to implement Orchard validation for transaction v5 (#2379)

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

As part of NU5, there is a [new transaction format and consensus rules (version 5)](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus). Zebra already had initial support for the new transaction, but the consensus rules aren't fully implemented.

This is the sixth and final PR that's part of #1981, and therefore this PR closes #1981.

### Specifications

The new transaction encoding and consensus rules are detailed in the [specification](https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus).

## Solution

The `verify_v5_transaction` method now also includes the checks generated by calling the `verify_sapling_shielded_data` method.

Only one test is included, even though there are two tests for V4 transactions. Unfortunately, the test where there are no Sapling spends but at least one Sapling output can't be implemented because there are no fake V5 transactions that satisfy this condition. The transactions that satisfy this for V4 have Sprout joinsplits as inputs, which are then removed when converted into a V5 transaction, making the verifier reject them due to missing inputs.

## Review

@teor2345 has already reviewed the other PRs that were part of #1981.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

**Comment by teor2345** [comment]
I'd like @oxarbitrage to do the initial review on this one, with @dconnolly's help if needed.

**Comment by jvff** [comment]
Added a `#[should_panic]` to skip a failing test, because it currently requires the sighash implementation for V5 transactions. That can be removed once #2165 is merged.

File: zebra-consensus/src/transaction/tests.rs [hunk 1]
```diff
@@ -201,6 +201,8 @@ fn v5_coinbase_transaction_with_enable_spends_flag_fails_validation() {
 }
 
 #[tokio::test]
+// TODO: Remove `should_panic` once V5 transaction sighash is implemened.
```
Comment (oxarbitrage) [inline zebra-consensus/src/transaction/tests.rs]:
```suggestion
// TODO: Remove `should_panic` once V5 transaction sighash is implemened by the merge of #2165.
```

File: zebra-consensus/src/transaction/tests.rs [hunk 2]
```diff
@@ -775,6 +777,47 @@ fn v4_with_sapling_outputs_and_no_spends() {
     });
 }
 
+/// Test if a V5 transaction with Sapling spends is accepted by the verifier.
+#[test]
+// TODO: Remove `should_panic` once V5 transaction verification is complete.
```
Comment (oxarbitrage) [inline zebra-consensus/src/transaction/tests.rs:782]: do we have a ticket number or PR for this ?

**Review by oxarbitrage — COMMENTED** [review]
This looks good to me but i am unsure if it is fully closing https://github.com/ZcashFoundation/zebra/issues/1981

Can we make sure these are done in previous PRs or open additional tickets?

Other transparent and sapling consensus rules:
- [ ] quote the relevant consensus rules in each function's docs
- [ ] add orchard TODOs
- [ ] refactor to remove duplicate code, and make the consensus rule logic clearer

File: zebra-consensus/src/transaction/tests.rs (hunk 2 above)
Comment (jvff) [reply zebra-consensus/src/transaction/tests.rs:782]: I couldn't find an issue to track full V5 transaction support, but I did add a `TODO` to the main code in [77337c1](https://github.com/ZcashFoundation/zebra/pull/2437/commits/77337c1736be2b6ec9e2b90878f527291f4363c5).

**Comment by jvff** [comment]
> This looks good to me but i am unsure if it is fully closing #1981
> 
> Can we make sure these are done in previous PRs or open additional tickets?
> 
> Other transparent and sapling consensus rules:
> 
>     * [ ]  quote the relevant consensus rules in each function's docs
> 
>     * [ ]  add orchard TODOs
> 
>     * [ ]  refactor to remove duplicate code, and make the consensus rule logic clearer

I'm not exactly sure if they all have been implemented, but here's my view on the items:

- **quote the relevant consensus rules in each function's docs**: these are quoted and linked to inside the function's comments. Since #1981 tracks transparent and Sapling validation in V5 transactions, AFAICT no new consensus rules are directly implemented, the same rules are reused for V5. But if the plan is to move the quotes into actual documentation, it might make sense to refactor the methods to make it easier to read the rules and the code, perhaps even moving each rule into a function inside the `check` module. Any opinions? Maybe I should open an issue to discuss this?
- **add orchard TODOs**: some were already there, but I did add some more and link to some issues.
- **refactor to remove duplicate code, and make the consensus rule logic clearer**: I think this is done. I mean, the code is now shared and there's no little to no duplicate code. However, I guess there will always be room for improvement :laughing:

**Review by oxarbitrage — APPROVED** [review]

File: zebra-consensus/src/transaction/tests.rs (hunk 2 above)
Comment (teor2345) [reply zebra-consensus/src/transaction/tests.rs:782]: > I couldn't find an issue to track full V5 transaction support

FYI that's #2379 - Validate remaining Orchard consensus rules in V5 transactions.

**Comment by teor2345** [comment]
> * But if the plan is to move the quotes into actual documentation, it might make sense to refactor the methods to make it easier to read the rules and the code, perhaps even moving each rule into a function inside the `check` module. Any opinions? Maybe I should open an issue to discuss this?

I think opening an issue is a good idea.

But let's save any work on the ticket until after NU5 testnet activation - it can wait until it gets scheduled in a sprint.

**Pull request merged**